cpp/util/ManagedArray.h

```diff
@@ -154,67 +152,80 @@ template<typename T> class ManagedArray
 
     //*************************************************************************
     // In order to support convenient indexing using arbitrary numbers of
-    // indices, we provide overloads of the indexing operator. For convenience,
-    // all calls eventually funnel through the simplest function interface, a
-    // std::vector of indices. However, the more convenient approach is enabled
-    // using variadic template arguments.
+    // indices, we provide overloads of the indexing operator.  All calls
+    // eventually funnel through the simplest function interface, a std::vector
+    // of indices. However, the more convenient approach is enabled using
+    // variadic template arguments.
+    //
+    // Performance note: the variadic indexers can be expensive
+    // since they have to build up a std::vector through a sequence of
+    // (hopefully compiler optimized) function calls. In general, this is
+    // unimportant since performance critical code paths in freud do not
+    // involve large amounts of indexing into ManagedArrays. However, in cases
+    // where critical code paths do index into the array, it may be
+    // advantageous to directly call the function with a std::vector of indices
+    // and bypass the variadic indexers.
```

Comment:
Given the performance constraints we're seeing on variadic indexing, I am somewhat inclined to remove variadic indexing and just supply `operator()` overloads for 1, 2, 3, and 4-dimensional ManagedArrays (similar to the `IndexND` approach of the previous codebase) and require vector notation for anything higher. The operator overload would become considerably simpler and more maintainable (though more repetitive), and it wouldn't cause these performance issues (as far as I can tell):
```c++
// 4D case
inline T &operator()(int i, int j, int k, int l)
{
    return (*this)({i, j, k, l});
}
```

I would include up to 4D so that we can handle cubatic tensors, and that's the highest dimensionality I'm aware of in freud.

Reply:
Just a suggestion -- if the problem is adequately solved and you don't see the variadic slowdown being a problem elsewhere, then we can leave this as-is. But the comments you wrote seem to indicate this is a significant slowdown, so I thought I'd pose an alternative.

Reply:
I don't think there's a strong reason overall to prefer either approach (there are pros and cons to both with respect to complexity vs duplication/generality), but I do think the problem is adequately solved. In general, Steinhardt is the first time that I'm finding this problem, and it's specifically because in `computeAve` we use the indexer inside like 4 nested for loops _and_ the indexing is the only operation in there. I haven't noticed the problem in any other changes. We can definitely revisit this if we start seeing the slowdown more often, but for now I don't think it's worth overreacting to this one use-case.

I would consider this difference to be similar to the difference between directly indexing into vs use `at` for `std::vector`, there are tradeoffs to both approaches and I think it's OK to just have the options available.